feat(chat): Implement multi-modal image input

lars-hagen opened the pull request:
This PR introduces the ability for users to attach and send images along with text messages in the chat interface. 
**Note: This initial implementation adds multi-modal support specifically for the OpenAI provider.**

Key Changes:

Frontend (`Input.tsx`, `ChatView.tsx`, `UserMessage.tsx`):
*   Allows pasting images into the input field.
*   Displays image previews below the input text area.
*   Allows removing attached images before sending.
*   Sends both text and image data (as base64 URIs) in the `createUserMessage` payload.
*   Renders user-uploaded images within the `UserMessage` component.
*   Updates `handleSubmit` and `createUserMessage` to handle the new multi-modal structure.

Backend (`openai.rs`, `utils.rs`):
*   Refactors the `format_messages` function for the OpenAI provider to correctly handle the new array-based content format (mixing text and image parts in a single message).
*   Ensures tool responses containing images are handled gracefully (placeholder text + separate image message).
*   Updates `convert_image` utility to correctly format data URIs for OpenAI.

Here is a screenshot of how it looks:

![1-96f09dc5](https://github.com/user-attachments/assets/8090fb78-9a08-4472-93c5-fbeedfc529ad)

nahiyankhan commented:
@lars-hagen awesome! appreciate this work. I am attaching designs that our designer had around this. note - there are concepts in the screens thats not implemented yet. but key things to look at are - the drag and drop support (this could be split into a separate pr) and the image preview style. 

<img width="976" alt="Screenshot 2025-04-10 at 8 28 17 PM" src="https://github.com/user-attachments/assets/040a61d1-64d9-462b-b877-974db0bf5665" />

other than that, if this is openai specific, we likely have to handle the cases for when its not 🤔

dantesutton commented on crates/goose/src/providers/formats/openai.rs:
```diff
@@ -173,19 +170,38 @@ pub fn format_messages(messages: &[Message], image_format: &ImageFormat) -> Vec<
+        let mut should_add = false;
+
+        // If there was text or image content, add it as an array
+        if has_content {
```
Possible to use `current_content_parts.is_empty()` instead of `has_content`? `has_content = true` always follows a push, so it doesn't seem needed at the moment.

Were there any other use cases for `has_content` that you were thinking of?

dantesutton commented on crates/goose/src/providers/formats/openai.rs:
```diff
@@ -93,77 +91,76 @@ pub fn format_messages(messages: &[Message], image_format: &ImageFormat) -> Vec<
                             }));
                         }
                     }
+<<<<<<< HEAD
```
potential accident from merge conflict?

lars-hagen replied:
Thanks for spotting this. Will be fixed in latest commit.

dantesutton commented on crates/goose/src/providers/formats/openai.rs:
```diff
@@ -173,19 +170,38 @@ pub fn format_messages(messages: &[Message], image_format: &ImageFormat) -> Vec<
-        if converted.get("content").is_some() || converted.get("tool_calls").is_some() {
-            output.insert(0, converted);
+        // After processing all content parts for the message:
+        let mut message_to_add = converted; // Start with role
```
nit: qq, why do we create the converted variable at the top of the file if we can just do the following?

`let mut message_to_add = json!({
            "role": message.role
        });
`

lmk if im missing any other context of this `converted` variable

lars-hagen replied:
Will be fixed in latest commit.

lars-hagen commented:
Adressing the feedback from @nahiyankhan and @spencrmartin and various other improvements.

**Drop Zone UI:**
*   Added an `Upload` icon to visually indicate the drop target area.
*   Updated the drop zone styling to use a neutral grey dashed border and removed the background color.
*   Refined the indicator text to be more descriptive: "Drop files here to upload into your goose chat".

**Image Preview Styling:**
*   Reduced the size of attached image previews in the input area.
*   Repositioned the "remove" (X) button to the top-right corner of the preview and adjusted its offset for better spacing.

These changes aim to improve the drag-and-drop user experience and align the image preview styling closer to the design feedback.

Here is a video of how it looks.

https://github.com/user-attachments/assets/f4544de4-35fc-4ed2-a097-77aa08926b96

michaelneale commented:
woah this is cool! so for non openai providers - what will happen, won't behave the same but won't break?

lily-de commented on crates/goose/src/providers/formats/openai.rs:
```diff
@@ -17,51 +17,40 @@ use serde_json::{json, Value};
-                    // Redacted thinking blocks are not directly used in OpenAI format
-                    continue;
-                }
+                MessageContent::Thinking(_) => continue,
+                MessageContent::RedactedThinking(_) => continue,
```
can we get those comments back in?

lily-de commented on ui/desktop/src/components/UserMessage.tsx:
```diff
@@ -12,26 +12,47 @@ interface UserMessageProps {
+                ref={index === 0 ? contentRef : null}
+                className="user-message"
+              >
+                <MarkdownContent content={part.text} className="text-white prose-a:text-white" />
```
```suggestion
                <MarkdownContent content={part.text} className="text-white prose-a:text-white user-message" />
```

taniashiba commented:
Some reviewed comments left for you, @lars-hagen ! Let us know if you have any additional followup questions. Thank you so much for your contribution <3

BjoernRave commented:
is there a way to provide images using the cli? maybe even through recipes?

michaelneale commented:
looks like this has been done in a different way in the meantime - is sthis still relevant?

lars-hagen commented:
No I'll close it down. Thanks and good job team goose on your implementation of the multi modal looks nice :)
